### PR TITLE
Fixed #28828 -- Improved performance of HttpRequest.build_absolute_uri().

### PR DESCRIPTION
**Pull request by gcbirzan**
A fix for https://code.djangoproject.com/ticket/28828.

In cases where a lot of URLs are generated, the performance of it is suboptimal. 
It calls self.get_host() several times. That function does a lot of work that isn't needed more than once.
It calls urljoin() even on the trivial case of absolute paths (which is the most common case). urljoin is quite expensive as well.
My patch fixes these by using a cached_property for the scheme://host part and by just concatenating the current scheme/host with the location when the location is absolute, doesn't change the host, and doesn't contain '.' or '..'. The last can be implemented in there, but that will just complicate the code I think, for again, a not very common case.

### CODE REVIEW

File: django/http/request.py [hunk 1]
```diff
@@ -176,15 +177,29 @@ def build_absolute_uri(self, location=None):
...
         bits = urlsplit(location)
         if not (bits.scheme and bits.netloc):
-            current_uri = '{scheme}://{host}{path}'.format(scheme=self.scheme,
-                                                           host=self.get_host(),
-                                                           path=self.path)
-            # Join the constructed URL with the provided location, which will
-            # allow the provided ``location`` to apply query strings to the
-            # base path as well as override the host, if it begins with //
-            location = urljoin(current_uri, location)
+            # Handle the trivial case trivially. If the location is absolute
```
Comment (kezabelle) [inline django/http/request.py]: The trivial case is now less trivial, with this change :) [I get the jist of what you're saying, though]

File: django/http/request.py [hunk 2]
```diff
@@ -176,15 +177,29 @@ def build_absolute_uri(self, location=None):
...
+            # and no scheme or host (netloc) is provided we can shortcut
+            # urljoin as long as no path segments are '.' or '..'.
+            if (bits.path.startswith('/') and not bits.scheme and not bits.netloc and
```
Comment (kezabelle) [inline django/http/request.py]: If we're into optimising, I believe getting the first char by index is still faster than `startswith`, which this back-of-the-napkin test seems to agree with:
```
> python3.5 -m timeit 'y = "/this/is/a/test"; y.startswith("/")'
1000000 loops, best of 3: 0.266 usec per loop
> python3.5 -m timeit 'y = "/this/is/a/test"; y[0]=="/"'
10000000 loops, best of 3: 0.0917 usec per loop
```
I don't know if its *worth* it, but there's another option.

File: django/http/request.py [hunk 3]
```diff
@@ -176,15 +177,29 @@ def build_absolute_uri(self, location=None):
...
+                    not any(seg == '.' or seg == '..' for seg in bits.path.split('/'))):
+                # If location starts with '//' but has no netloc that means we
+                # should reuse the current one. Strip the double slashes and
```
Comment (kezabelle) [inline django/http/request.py]: Does `current one` means *scheme relative* or *host relative* here? Looking at the test, I think its the latter, but I'd originally guessed the former.

File: django/http/request.py (hunk 2 above)
Comment (gcbirzan) [reply django/http/request.py]: Good point. I'll fix that.

File: django/http/request.py (hunk 3 above)
Comment (gcbirzan) [reply django/http/request.py]: Current one means the one only one we have, the one from the current request.

File: django/http/request.py (hunk 2 above)
Comment (gcbirzan) [reply django/http/request.py]: Actually, not entirely sure this should be fixed here. The semantics are correct, it does read well, it's just that cpython is terrible at it.

The changes I suggest have a way bigger impact than this and if you (not necessarily you personally) don't disagree, I think readability should go first. As it stands, I hate this bit since it's hard to follow, adding that change will make it worse in that regard, for a very minor gain.

File: django/http/request.py (hunk 1 above)
Comment (gcbirzan) [reply django/http/request.py]: Unfortunately, I spent probably half the time on this PR fixing the comments. If you have a better wording, please let me know. "Shortcut the trivial case" is the other one I came up with, but I liked the other one better.

File: django/http/request.py (hunk 2 above)
Comment (gcbirzan) [reply django/http/request.py]: Actually. I'll do some benchmarking tomorrow to see the effect of each of these.

File: django/http/request.py (hunk 1 above)
Comment (kezabelle) [reply django/http/request.py]: The wording *I* would use is probably `Handle the common case: If the location ...` as the suggestion is that a) it happens far more frequently b) its not the pathological case (being the one which calls `urljoin`, now)
I'm not going to endlessly bikeshed over it. It's basically fine and does its job as is.

File: django/http/request.py (hunk 2 above)
Comment (kezabelle) [reply django/http/request.py]: > The changes I suggest have a way bigger impact than this and if you (not necessarily you personally) don't disagree, I think readability should go first.

I don't disagree - it is Python after all :)
I merely offered it up as something additional, as it's unclear from reading the code how much difference the change *as is already implemented* is making (and the related DRF ticket only gives an overview of where time is spent as far as I can see)
If it doesn't nudge the needle of whatever benchmarks you're using to check this is faster, it's probably not worth it.

If you have any numbers to demonstrate the difference this patch brings, that'd be good to see (either on the ticket or on the PR)

Comment (gcbirzan) [reply django/http/request.py]: I ran this benchmark:

    timeit.timeit("request.build_absolute_uri(location='///foo/bar/')", number=1000000, globals={'request': request})
The results were:
* With my fix: 4.474267777055502
* With my fix but with bits.path[0] == '/': 4.34382488578558
* My shortcut but no cached property: 9.473239112645388 [this is to simulate running this on different requests)
* No shortcut but with cached property: 12.506602805107832 
* Original version: 17.600296460092068

There is a regression. Running without the cached property, to simulate calling it in different requests, on a path with '.':

    timeit.timeit("request.build_absolute_uri(location='/foo/./bar/')", number=1000000, globals={'request': request}))

The results:
* My version, no cached property: 19.713809736073017
* Original version: 18.129451751708984

So, the extra checks do add some overhead, but it's for an uncommon case. I don't have any evidence, obviously, but I'm fairly confident that the overwhelming uses build_absolute_url() are with an absolute path that doesn't have '.' or '..' in it and that the result is coming from reverse.

Comment (gcbirzan) [reply django/http/request.py]: After the suggestion by @kezabelle to replace the check for . and .. with one that doesn't use split, the new time it takes to run without the cached property for /foo/./bar/ is 18.46866210922599, bringing it within 2% of the original.

File: django/http/request.py [hunk 4]
```diff
@@ -176,15 +177,29 @@ def build_absolute_uri(self, location=None):
...
-                                                           host=self.get_host(),
-                                                           path=self.path)
-            # Join the constructed URL with the provided location, which will
-            # allow the provided ``location`` to apply query strings to the
-            # base path as well as override the host, if it begins with //
-            location = urljoin(current_uri, location)
+            # Handle the most common and trivial case trivially. If the location is
+            # absolute and no scheme or host (netloc) is provided we can shortcut
+            # urljoin as long as no path segments are '.' or '..'.
+            if (bits.path.startswith('/') and not bits.scheme and not bits.netloc and
```
Comment (jaap3) [inline django/http/request.py]: No need to check for `bits.scheme` and `bits.netloc` again as this block is fenced by `if not (bits.scheme and bits.netloc)`

File: django/http/request.py [hunk 5]
```diff
@@ -176,15 +177,29 @@ def build_absolute_uri(self, location=None):
...
+                    '/./' not in bits.path and '/../' not in bits.path):
```
Comment (jaap3) [inline django/http/request.py]: This check doesn't catch a location that ends with `/.` or `/..`

File: django/http/request.py [hunk 6]
```diff
@@ -176,15 +177,29 @@ def build_absolute_uri(self, location=None):
...
+                # If location starts with '//' but has no netloc that means we
+                # should reuse the schema and netloc from the current request.
+                # Strip the double slashes and continue as if it wasn't specified.
+                if location.startswith('//'):
```
Comment (jaap3) [inline django/http/request.py]: I don't get this, if location is `//foo` then `bits.netloc` will be `foo`. The only way to get here is if location is prefixed by 4 (or more) forward slashes i.e. `////foo` or `/////foo` etc. which seems like uncommon input and should probably just be handled by `urljoin`

Comment (jaap3) [reply django/http/request.py]: `if (bits.path.startswith('/') and not bits.scheme and not bits.netloc and '/./' not in bits.path and '/../' not in bits.path):`

Could potentially be rewritten to:

```
path_bits = bits.path.split('/')
if (path_bits[0] == '' and not {'.', '..'}.intersection(path_bits)):
```

File: django/http/request.py (hunk 5 above)
Comment (gcbirzan) [reply django/http/request.py]: Correct, I might need to revert that bit.

File: django/http/request.py (hunk 4 above)
Comment (gcbirzan) [reply django/http/request.py]: `not (bits.scheme and bits.netloc)` is `not bits.scheme or not bits.netloc`, I want to check if none of them is true.

File: django/http/request.py (hunk 6 above)
Comment (gcbirzan) [reply django/http/request.py]: If location starts with `//` and there is no netloc and no schema, that means the original location started with `///`, we can just strip that since it means "same scheme, same host"

File: django/http/request.py (hunk 5 above)
Comment (kezabelle) [reply django/http/request.py]: I knew it'd miss that when I threw it out there, but I assume the slow path (going through `urljoin`) handles that. 
IMHO it's yet another edge case, and a harder one to catch (I think the quickest way to do it would be `bits.path[-3:] == "/.."` and `bits.path[-2:] == '/.'` maybe?)

**Comment by timgraham** [comment]
Additional test coverage should be added in a separate commit so that it's clear that the refactoring doesn't change behavior. I've included the additional tests here as part of https://github.com/django/django/pull/9566.
